proposed/2022/PrivateAssetIndepdentFromExcludeAsset.md

```diff
@@ -0,0 +1,147 @@
+
+## Motivation
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+A package author should be able control which asset flow to parent consuming projects when this project is consumed as package. It enables parent projects to consume transitive packages without having to directly reference them, therefore it reduces the number of packages developers need to keep track of.
+We couldn't make this default experience because it'll break customers who rely on current behavior.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+The new ` PackPrivateAssetsFlow` property only affects pack operation (more specifically nuspec in nupkg file), but doesn't affect restore experience for current project so there would be no change in `project.assets.json` lock file.
+Experience for parent consuming project would be affected by which asset from `compile, runtime, contentFiles, build, buildMultitargeting, buildTransitive, analyzers, native` are flowing into them.
+
+For the following table assume `PackPrivateAssetsFlow` is set `true` when creating package, iterating possible scenarios (not full list) for consuming parent project.
+
+| Asset flowing to parent project | New feature enabled | Possible downside |
+|-----------------------|--------------|-----------------|
+| compile | Expose new code, auto complete, intellicode, intellisense | Compile error due to naming ambiguity, run time error, NU1605 error |
+| runtime | Let provide new runtime dependency | Run time error |
+| build | Enable msbuild imports | Build fails due to property/target value change |
+| analyzers | Code analyzers work | n/a |
+
+#### Examples
+
+##### Case 1
+
+Package reference in csproj file.
+
+```.net
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <VersionSuffix>beta</VersionSuffix>
+    <PackPrivateAssetsFlow>True</PackPrivateAssetsFlow>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Windows.CsWin32" Version="0.2.138-beta" PrivateAssets="none" ExcludeAssets="build" />
+  </ItemGroup>
+
+Before change nuspec file:
+
+```.net
+    <dependencies>
+      <group targetFramework=".NETStandard2.0">
+        <dependency id="Microsoft.Windows.CsWin32" version="0.2.138-beta" exclude="Runtime,Build,Native,Analyzers,BuildTransitive" />
+      </group>
+    </dependencies>
+```
+
+After change nuspec file:
+
+```.net
+    <dependencies>
+      <group targetFramework=".NETStandard2.0">
+        <dependency id="Microsoft.Windows.CsWin32" version="0.2.138-beta" include="All" />
```

Comment:
The location of these docs is obscure, but this doc says: https://learn.microsoft.com/en-us/nuget/create-packages/select-assemblies-referenced-by-projects

>  Note that the default private assets when none is supplied is `contentfiles;build;analyzers`

(although I think it might be a bug in the docs, I don't think that `analyzers` are in the default private assets, but I'm not confident).

This proposal is changing the defaults. While it could be argued that it's not technically a breaking change, I'm concerned about it.

The design of PackageReference is that `build` is for direct references only, and `buildTransitive` is for both direct and transitive packages. Removing `build` from the default exclude might cause problems, and removing `contentFiles` from the default exclude might cause unwanted content to start being included in package consumers projects.

Reply:
analyzers are in the default, they're just not implemented yet.

Reply:
Changing the default is not a goal, @zivkan. In this example, `include="All"` is present in the nuspec file because `PrivateAssets="none"` is specified on the PackageReference, which overrides any default.

Reply:
Ok, thanks. I think I'm starting to understand the intended outcome for the feature.